### ISSUE TITLE
Too slow serialization of ADT-based Scala enums

### ISSUE THREAD

**Issue by plokhotnyuk** [ngs-doo/dsl-json#157]
Currently parsing of ADT-based Scala enums with DSL_JSON is more than 700x times faster than serialization.

Please see results of [benchmarks](https://github.com/plokhotnyuk/jsoniter-scala/tree/master/jsoniter-scala-benchmark/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/benchmark) from JDK 8 which test parsing of an array of 128 enums by different JSON parsers for Scala: 
```
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                     (size)   Mode  Cnt        Score      Error  Units
[info] ArrayOfEnumADTsReading.avSystemGenCodec          128  thrpt    5   138077.169 ±  560.317  ops/s
[info] ArrayOfEnumADTsReading.borerJson                 128  thrpt    5   148858.068 ± 1437.148  ops/s
[info] ArrayOfEnumADTsReading.circe                     128  thrpt    5    56051.325 ±  474.688  ops/s
[info] ArrayOfEnumADTsReading.dslJsonScala              128  thrpt    5   294570.386 ± 2348.747  ops/s
[info] ArrayOfEnumADTsReading.jacksonScala              128  thrpt    5   160523.150 ± 1851.129  ops/s
[info] ArrayOfEnumADTsReading.jsoniterScala             128  thrpt    5   343154.537 ± 1681.139  ops/s
[info] ArrayOfEnumADTsReading.playJson                  128  thrpt    5    71647.259 ±  255.676  ops/s
[info] ArrayOfEnumADTsReading.sprayJson                 128  thrpt    5    76016.362 ±  307.266  ops/s
[info] ArrayOfEnumADTsReading.uPickle                   128  thrpt    5   116074.985 ±  680.481  ops/s
[info] ArrayOfEnumADTsWriting.avSystemGenCodec          128  thrpt    5   127695.961 ±  186.634  ops/s
[info] ArrayOfEnumADTsWriting.borerJson                 128  thrpt    5   220077.216 ±  971.548  ops/s
[info] ArrayOfEnumADTsWriting.circe                     128  thrpt    5    33194.108 ±  267.881  ops/s
[info] ArrayOfEnumADTsWriting.dslJsonScala              128  thrpt    5      399.149 ±    9.338  ops/s
[info] ArrayOfEnumADTsWriting.jacksonScala              128  thrpt    5   300119.810 ± 1185.247  ops/s
[info] ArrayOfEnumADTsWriting.jsoniterScala             128  thrpt    5  1045616.683 ± 4158.261  ops/s
[info] ArrayOfEnumADTsWriting.jsoniterScalaPrealloc     128  thrpt    5  1102949.348 ± 2048.285  ops/s
[info] ArrayOfEnumADTsWriting.playJson                  128  thrpt    5   251161.224 ± 1633.918  ops/s
[info] ArrayOfEnumADTsWriting.sprayJson                 128  thrpt    5   121418.777 ±  900.754  ops/s
[info] ArrayOfEnumADTsWriting.uPickle                   128  thrpt    5    36972.850 ±  273.809  ops/s
```
To reproduce results you can clone [the jsoniter-scala repo](https://github.com/plokhotnyuk/jsoniter-scala) and run the following command:
```sh
sbt 'jsoniter-scala-benchmark/jmh:run -p size=128 ArrayOfEnumADTs'
```

Sources of the problem can be easily spotted on the flame graph for CPU usage. Please download an attached archive of the SVG file and open it the browser to explore interactively.

[flame-graph-cpu.zip](https://github.com/ngs-doo/dsl-json/files/4075845/flame-graph-cpu.zip)

Also, you can build it for yourself after cloning of [async-profiler](https://github.com/jvm-profiling-tools/async-profiler) and [FlameGraph](https://github.com/brendangregg/FlameGraph) repos and using a command like:
```sh
sbt 'jsoniter-scala-benchmark/jmh:run -p size=128 -jvmArgsAppend "-XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints" -prof "jmh.extras.Async:event=cpu;interval=1234567;dir=/tmp/profile-async;asyncProfilerDir=/home/andriy/Projects/com/github/jvm-profiling-tools/async-profiler;flameGraphDir=/home/andriy/Projects/com/github/brendangregg/FlameGraph;flameGraphOpts=--color,java;verbose=true" -wi 10 -i 50 ArrayOfEnumADTsWriting.dslJsonScala'
```

**Comment by zapov** [comment]
I'll try to look into later, but if I'm reading this graphs correctly it's constantly trying to lookupFromFactories to create a specialized encoder. It should be doing that only once and then reuse the cached version on later calls.

So there is some problem with your setup which causes repeated calls to analysis ;(

**Comment by plokhotnyuk** [comment]
Setup of the benchmark is quite straightforward it uses `dslJson.encoder` and `dslJson.decoder` call to create `JsonWriter.WriteObject` and `JsonReader.ReadObject` instances accordingly, store them in local values. Instances of `JsonWriter` and `JsonReader` are stored too as thread local values. Then for parsing and serialization the following functions are used:
```
  def dslJsonDecode[T](bytes: Array[Byte])(implicit decoder: JsonReader.ReadObject[T]): T = {
    val reader = threadLocalJsonReader.get().process(bytes, bytes.length)
    reader.read()
    decoder.read(reader)
  }

  def dslJsonEncode[T](obj: T)(implicit encoder: JsonWriter.WriteObject[T]): Array[Byte] = {
    val writer = threadLocalJsonWriter.get()
    writer.reset()
    encoder.write(writer, obj)
    writer.toByteArray
  }
```
Full sources of the setup for DSL-JSON are [here](https://github.com/plokhotnyuk/jsoniter-scala/blob/master/jsoniter-scala-benchmark/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/benchmark/DslPlatformJson.scala). It will be greatly appreciated if you will look through them and give your feedback of how it can be improved.

**Comment by zapov** [comment]
I see where the problem is. Tnx for the report. Will fix it in next version.

For now as a workaround you can explicitly register the converter as:

    dslJson.registerWriter(classOf[Array[ADT]], new ArrayEncoder(dslJson, dslJson.encoder[ADT]))

**Comment by plokhotnyuk** [comment]
@zapov Thank you a lot! 

The proposed W/A works greatly for me:
```
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                            (size)   Mode  Cnt       Score      Error  Units
[info] ArrayOfEnumADTsReading.dslJsonScala     128  thrpt    5  287046.119 ± 2299.241  ops/s
[info] ArrayOfEnumADTsWriting.dslJsonScala     128  thrpt    5  512861.445 ± 2207.656  ops/s
```

**Comment by zapov** [comment]
btw. on Windows 7 I'm unable to run the sbt commands for your project. Even starting the sbt does not work. When I kill it I get 
```[info] Loading settings for project global-plugins from idea.sbt ...
2020-01-17 15:15:51,714 shutdown-hooks-run-all ERROR No Log4j 2 configuration fi
le found. Using default configuration (logging only errors to the console), or u
ser programmatically provided configurations. Set system property 'log4j2.debug'
 to show Log4j 2 internal initialization logging. See https://logging.apache.org
/log4j/2.x/manual/configuration.html for instructions on how to configure Log4j
2

**Comment by plokhotnyuk** [comment]
@zapov Thanks for the feedback!

I've never tried to build it on Windows 7 locally, but [these tunings](https://github.com/plokhotnyuk/jsoniter-scala/blob/master/.github/workflows/build.yml#L33-L34) of the Github Actions setup was required to run it properly for the latest version of Windows.

**Comment by zapov** [comment]
Fixed in v1.9.6.
v1.9.6 released
